### DISCUSSION TITLE
Explicit CHANGELOG.md display

### DISCUSSION THREAD

**Discussion by calimarkus** [SwiftPackageIndex/SwiftPackageIndex-Server#1857] [Ideas]
Currently the SPI has a prominent [tab for release notes](https://swiftpackageindex.com/calimarkus/JDStatusBarNotification#releases). I assume that's powered by GH releases, right?
Would you consider exposing `CHANGELOG.md` files (if existing) as a tab (or otherwise), similar to what [CocoaPods does](https://cocoapods.org/pods/JDStatusBarNotification#changelog)?

PS: I didn't have any official releases tagged on GH. Thus the tab was fully empty. Should it be hidden in that case?

**SPI:**

[<img width="502" alt="Screen Shot 2022-06-29 at 12 05 14 PM" src="https://user-images.githubusercontent.com/807039/176343310-9ed90da3-c23f-4ad4-8fd9-60177354d34c.png">](https://swiftpackageindex.com/calimarkus/JDStatusBarNotification#readme)

**CocoaPods:**

[<img width="369" alt="Screen Shot 2022-06-29 at 12 05 26 PM" src="https://user-images.githubusercontent.com/807039/176343405-c3935467-6382-424a-b1dc-33dedb668579.png">](https://cocoapods.org/pods/JDStatusBarNotification#changelog)

**Comment by daveverwer** [comment]
> Currently the SPI has a prominent [tab for release notes](https://swiftpackageindex.com/calimarkus/JDStatusBarNotification#releases). I assume that's powered by GH releases, right? 

That's correct, yes.

> PS: I didn't have any official releases tagged on GH. Thus the tab was fully empty. Should it be hidden in that case?

Possibly. We've also talked about formatting commit messages into that tab, a little like GitHub also does. @Sherlouk even had a prototype of that working but we decided it needed to be a little richer in its display, which makes it a bigger feature to undertake.

> Would you consider exposing `CHANGELOG.md` files (if existing) as a tab (or otherwise), similar to what [CocoaPods does](https://cocoapods.org/pods/JDStatusBarNotification#changelog)?

Certainly it's an interesting idea. I'd be interested to know what percentage of packages use a Changelog before committing to it.

Would you have an interest in working on the feature if that research turned up good numbers?

**Comment by calimarkus** [comment]
> Would you have an interest in working on the feature if that research turned up good numbers?

Possibly yeah!

**Comment by daveverwer** [comment]
Interesting! Here are some very brief thoughts on what I think will be involved. It's not going to be a trivial feature to implement but is certainly possible.

The first thing we'll need to do is figure out which packages have a Changelog. The best place to do this will probably be the analysis process (which has access to a checked-out version of the repository) and we'll want to store a boolean value in the `versions` table. We'll probably never use any version record apart from the default branch version record, but we'll need a field in the database so it will need to be there for all. This work should be merged as a PR once completed so that we can run a full re-analysis cycle ahead of time so that the data is all there when the display code goes live.

Then, we'll need to display the changelog for pages where that flag and display a new tab. I don't think it should replace the Release Notes tab as it's quite possible repositories will have both.

In terms of rendering the changelog. I think we should leverage GitHub's rendering as we do for Readme files. They sanitise the HTML and render GFM tags which people may also be doing with their changelogs. We currently scrape the rendered GitHub page, which is not ideal at all (and we have #1095 logged to fix that) and it may make sense to do the same here, but we should check if the API mentioned in #1095 supports Changelog files, too, and if it does, consider prioritising #1095 before implementing this.

If you are interested, we'll need you to sign a standard Apache Contributor License Agreement before starting work. The easiest way to get that organised would be on [our Discord](https://discord.gg/vQRb6KkYRw).

**Comment by calimarkus** [comment]
There's the generic [`/markdown` rendering API](https://docs.github.com/en/rest/markdown) too, which could work for both cases. The Sublime Plugin [`facelessuser/MarkdownPreview`](https://github.com/facelessuser/MarkdownPreview) uses this for previewing markdown.

`/markdown` has a limit of 400kb for the file size though. It seems like [`/repos/{owner}/{repo}/readme`](https://docs.github.com/en/rest/repos/contents#get-a-repository-readme) doesn't have that limitation. Not sure what percentage of repos would go over that limit and if that would be acceptable. We could also try to see what CocoaPods does, if that part is OpenSource.

To sum up your suggestions:

- Decide if #1095 should be done first
- Update analysis process to store a boolean value in the `versions` table (requires analysis cycle run)
- Display a new tab, when flag is set, render using GH APIs

**Reply by daveverwer** [reply]
I'm sure there are a few things we've forgotten, but yes that sounds good!
